File: src/lib/tauri-api/checkpoints.ts
import type {
  WorkspaceCheckpoint,
  WorkspaceCheckpointBranchResult,
  WorkspaceCheckpointDeleteResult,
  WorkspaceCheckpointDiff,
  WorkspaceCheckpointRestorePlan,
  WorkspaceCheckpointRestoreResult,
} from '../../types/checkpoint';
import { invokeCommand } from './client';

export function listWorkspaceCheckpoints(workspaceId: string): Promise<WorkspaceCheckpoint[]> {
  return invokeCommand<WorkspaceCheckpoint[]>('list_workspace_checkpoints', { workspaceId });
}

export function createWorkspaceCheckpoint(workspaceId: string, reason?: string): Promise<WorkspaceCheckpoint | null> {
  return invokeCommand<WorkspaceCheckpoint | null>('create_workspace_checkpoint', { workspaceId, reason });
}

export function getWorkspaceCheckpointDiff(workspaceId: string, reference: string): Promise<WorkspaceCheckpointDiff> {
  return invokeCommand<WorkspaceCheckpointDiff>('get_workspace_checkpoint_diff', { workspaceId, reference });
}

export function getWorkspaceCheckpointRestorePlan(workspaceId: string, reference: string): Promise<WorkspaceCheckpointRestorePlan> {
  return invokeCommand<WorkspaceCheckpointRestorePlan>('get_workspace_checkpoint_restore_plan', { workspaceId, reference });
}

export function restoreWorkspaceCheckpoint(workspaceId: string, reference: string): Promise<WorkspaceCheckpointRestoreResult> {
  return invokeCommand<WorkspaceCheckpointRestoreResult>('restore_workspace_checkpoint', { workspaceId, reference });
}

export function deleteWorkspaceCheckpoint(workspaceId: string, reference: string): Promise<WorkspaceCheckpointDeleteResult> {
  return invokeCommand<WorkspaceCheckpointDeleteResult>('delete_workspace_checkpoint', { workspaceId, reference });
}

export function createBranchFromWorkspaceCheckpoint(
  workspaceId: string,
  reference: string,
  branch: string,
): Promise<WorkspaceCheckpointBranchResult> {
  return invokeCommand<WorkspaceCheckpointBranchResult>('create_branch_from_workspace_checkpoint', { workspaceId, reference, branch });
}

File: src-tauri/src/commands/checkpoints.rs
use tauri::State;

use crate::models::{
    WorkspaceCheckpoint, WorkspaceCheckpointBranchResult, WorkspaceCheckpointDeleteResult,
    WorkspaceCheckpointDiff, WorkspaceCheckpointRestorePlan, WorkspaceCheckpointRestoreResult,
};
use crate::services::checkpoint_service;
use crate::state::AppState;

#[tauri::command]
pub fn list_workspace_checkpoints(
    state: State<'_, AppState>,
    workspace_id: String,
) -> Result<Vec<WorkspaceCheckpoint>, String> {
    checkpoint_service::list_workspace_checkpoints(&state, &workspace_id)
}

#[tauri::command]
pub fn create_workspace_checkpoint(
    state: State<'_, AppState>,
    workspace_id: String,
    reason: Option<String>,
) -> Result<Option<WorkspaceCheckpoint>, String> {
    checkpoint_service::create_workspace_checkpoint(&state, &workspace_id, reason.as_deref())
}

#[tauri::command]
pub fn get_workspace_checkpoint_diff(
    state: State<'_, AppState>,
    workspace_id: String,
    reference: String,
) -> Result<WorkspaceCheckpointDiff, String> {
    checkpoint_service::get_workspace_checkpoint_diff(&state, &workspace_id, &reference)
}

#[tauri::command]
pub fn get_workspace_checkpoint_restore_plan(
    state: State<'_, AppState>,
    workspace_id: String,
    reference: String,
) -> Result<WorkspaceCheckpointRestorePlan, String> {
    checkpoint_service::get_workspace_checkpoint_restore_plan(&state, &workspace_id, &reference)
}

#[tauri::command]
pub fn restore_workspace_checkpoint(
    state: State<'_, AppState>,
    workspace_id: String,
    reference: String,
) -> Result<WorkspaceCheckpointRestoreResult, String> {
    checkpoint_service::restore_workspace_checkpoint(&state, &workspace_id, &reference)
}

#[tauri::command]
pub fn delete_workspace_checkpoint(
    state: State<'_, AppState>,
    workspace_id: String,
    reference: String,
) -> Result<WorkspaceCheckpointDeleteResult, String> {
    checkpoint_service::delete_workspace_checkpoint(&state, &workspace_id, &reference)
}

#[tauri::command]
pub fn create_branch_from_workspace_checkpoint(
    state: State<'_, AppState>,
    workspace_id: String,
    reference: String,
    branch: String,
) -> Result<WorkspaceCheckpointBranchResult, String> {
    checkpoint_service::create_branch_from_workspace_checkpoint(
        &state,
        &workspace_id,
        &reference,
        &branch,
    )
}
